server/models/dataset.py

```diff
@@ -327,16 +296,49 @@ def reviewImages(self, dataset, acceptedImages, flaggedImages, user):
                 Folder.countFolders(prereviewFolder)) == 0:
             Folder.remove(prereviewFolder)
 
-    def registerMetadata(self, dataset, csvFile):
+    def registerMetadata(self, dataset, csvFile, user, sendMail=False):
         """Register a .csv file containing metadata about images."""
-        # Add image metadata information to list
-        registeredMetadata = dataset['meta'].get('metadata', [])
-        registeredMetadata.append({
+        # Check if image metadata is already registered
+        if self.findOne({'meta.metadataFiles.fileId': csvFile['_id']}):
+            raise ValidationException('Metadata file is already registered.')
+
+        # Add image metadata file information to list
+        now = datetime.datetime.utcnow()
+        metadataFiles = dataset['meta'].get('metadataFiles', [])
+        metadataFiles.append({
             'fileId': csvFile['_id'],
-            'time': datetime.datetime.utcnow()
+            'user': user['_id'],
+            'time': now
         })
         dataset = self.setMetadata(dataset, {
-            'metadata': registeredMetadata
+            'metadataFiles': metadataFiles
         })
 
+        # Send email notification
+        if sendMail:
+            host = mail_utils.getEmailUrlPrefix()
+            self._mailGroup(
+                groupName='Dataset QC Reviewers',
+                templateFilename='registerMetadataNotification.mako',
+                templateParams={
+                    'host': host,
+                    'dataset': dataset,
+                    'user': user,
+                    'csvFile': csvFile,
+                    'date': now.replace(microsecond=0)
+                },
+                subject='ISIC Archive: Dataset Metadata Notification')
+
         return dataset
+
+    def _mailGroup(self, groupName, templateFilename, templateParams, subject):
```

Comment:
This is a nifty helper. It might be nice to move it somewhere more general. I don't know where yet.